File: src/dccex_usb/stream.py
"""The rule the monitor's stream rides on: what a browser sends, and what it
is sent.

The UI's monitor is a page, and a page cannot open a socket on 2560. What a
browser can open is a stream on the origin it is already on, which it does by
upgrading a request it has already made — a key on the way in, a token back,
and frames after that. The upgrade and the frames either side of it are this
file (ADR-0004 d.3, ADR-0007).

**The rule and nothing of the transport.** A handshake is a key answered with
a token, and a frame is a header, a mask and a payload. Nothing here holds a
socket, which is what lets the rule be read and tested on its own with no
mirror and no browser — `framing.py` is the same thing one protocol further
in.

**Every byte the station sends rides as binary.** A text frame has to carry
valid UTF-8 and a serial line promises none: one garbled byte would end the
stream, where the mirror's contract is that every byte reaches every client
unchanged (CONTEXT.md). What the page makes of them is the page's business.

**What a browser sends is payload and not a message.** Whole `<…>` messages
are the mirror's rule and the mirror is what applies it (`framing.py`), so a
frame's payload is handed on as the bytes it is — split across frames, or
several messages to a frame, the way a client's bytes arrive split across TCP
reads on 2560. Text and binary are the same bytes here; the browser chooses
which it types in.

**A client's frames are masked and a server's are not**, and a frame that
breaks the rule is not something to work around: it is a peer that is not
speaking the protocol, and the stream ends saying which rule it broke
(control ADR-0050).
"""

import base64
import hashlib
from typing import NamedTuple

GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
"""The constant a browser's key is answered with, so that the token can only
have come from something that understood the request (RFC 6455). It proves
nothing about who asked — what does that is the origin the face is holding the
page to (ADR-0004)."""

CONTINUATION = 0x0
TEXT = 0x1
BINARY = 0x2
CLOSE = 0x8
PING = 0x9
PONG = 0xA

DATA = (CONTINUATION, TEXT, BINARY)
"""The frames that carry a caller's bytes. All three are the same bytes to the
mirror: what ends a message is `>` and not a frame (`framing.py`)."""

CONTROL = (CLOSE, PING, PONG)
"""The frames that are about the stream rather than about the station."""

FIN = 0x80
RESERVED = 0x70
OPCODE = 0x0F
MASKED = 0x80
SIZE = 0x7F

TWO_BYTES = 126
EIGHT_BYTES = 127
MAX_CONTROL = 125
"""The longest a control frame may be, and the largest size a header carries
in its own byte."""

MAX_FRAME_BYTES = 1 << 16
"""How much of one frame this reads. A message the station answers to is at
most `framing.MAX_MESSAGE` bytes, so a frame of this size is already not a
command; reading a larger one would be a buffer somebody else decides the size
of (control ADR-0042)."""

NORMAL = 1000
GOING_AWAY = 1001
PROTOCOL_ERROR = 1002
TOO_LARGE = 1009
"""Why a stream ended, in the codes a browser reads: a goodbye, the end this
app is going to, a peer that is not speaking the protocol, and a frame larger
than is read here."""


class Unreadable(Exception):
    """What arrived is not a frame this reads.

    Carries the code the browser is told it on, because that is the whole of
    what is done about it: a peer that breaks the framing rule is reported and
    let go, never guessed at (control ADR-0050).
    """

    def __init__(self, code: int, reason: str) -> None:
        super().__init__(reason)
        self.code = code
        self.reason = reason


class Frame(NamedTuple):
    """One whole frame a browser sent: what kind it is, and its bytes
    unmasked."""

    opcode: int
    payload: bytes


def accepted(key: str) -> str:
    """The token that answers a browser's key."""
    return base64.b64encode(hashlib.sha1((key + GUID).encode()).digest()).decode()


def framed(payload: bytes, opcode: int = BINARY) -> bytes:
    """One frame out: whole, unmasked, and binary unless it is said otherwise.

    Whole because nothing here has a reason to fragment — what is being framed
    is one read off the device, already sized by the read.
    """
    size = len(payload)
    if size <= MAX_CONTROL:
        head = bytes((FIN | opcode, size))
    elif size < 1 << 16:
        head = bytes((FIN | opcode, TWO_BYTES)) + size.to_bytes(2, "big")
    else:
        head = bytes((FIN | opcode, EIGHT_BYTES)) + size.to_bytes(8, "big")
    return head + payload


def goodbye(code: int = NORMAL, reason: str = "") -> bytes:
    """The frame that ends a stream, and says why in a sentence a person can
    read out of a browser's console.

    The sentence is cut to what a control frame holds rather than making the
    frame bigger than the rule allows: what it carries is why, and a reason
    that would not fit is one nobody has written yet.
    """
    said = reason.encode()[: MAX_CONTROL - 2]
    return framed(code.to_bytes(2, "big") + said, CLOSE)


def unframed(buffered: bytes, arrived: bytes) -> tuple[bytes, list[Frame]]:
    """Fold `arrived` into `buffered` and take off every whole frame.

    Returns what is still a partial frame, to be passed back as `buffered`
    next time, and the frames that completed, in the order they did — the
    shape `framing.frame` has, for the same reason: bytes arrive in whatever
    chunks TCP hands over and a rule that needed them whole would be a rule
    about the network.

    Raises `Unreadable` where what arrived is not a frame this reads. A
    header is all it takes: a size larger than `MAX_FRAME_BYTES` is refused
    on sight rather than buffered up to, so a browser cannot name how much
    this holds for it.
    """
    rest = buffered + arrived
    frames: list[Frame] = []
    while True:
        frame, taken = _one(rest)
        if frame is None:
            return rest, frames
        frames.append(frame)
        rest = rest[taken:]


def _one(data: bytes) -> tuple[Frame | None, int]:
    """The first whole frame in `data` and how many bytes it took, or None
    where there is not a whole one yet."""
    if len(data) < 2:
        return None, 0
    first, second = data[0], data[1]
    opcode = first & OPCODE
    if first & RESERVED:
        raise Unreadable(PROTOCOL_ERROR, "a frame reserved bits are set on")
    if opcode not in DATA and opcode not in CONTROL:
        raise Unreadable(PROTOCOL_ERROR, f"a frame of a kind ({opcode}) nobody sends")
    if not second & MASKED:
        raise Unreadable(
            PROTOCOL_ERROR, "a client's frames are masked, and this one was not"
        )
    size = second & SIZE
    at = 2
    if size == TWO_BYTES:
        if len(data) < at + 2:
            return None, 0
        size = int.from_bytes(data[at : at + 2], "big")
        at += 2
    elif size == EIGHT_BYTES:
        if len(data) < at + 8:
            return None, 0
        size = int.from_bytes(data[at : at + 8], "big")
        at += 8
    if opcode in CONTROL and (size > MAX_CONTROL or not first & FIN):
        raise Unreadable(PROTOCOL_ERROR, "a control frame that is split or too long")
    if size > MAX_FRAME_BYTES:
        raise Unreadable(
            TOO_LARGE,
            f"a frame of {size} bytes is more than the stream reads"
            f" ({MAX_FRAME_BYTES})",
        )
    if len(data) < at + 4 + size:
        return None, 0
    mask = data[at : at + 4]
    at += 4
    payload = bytes(
        byte ^ mask[place % 4] for place, byte in enumerate(data[at:][:size])
    )
    return Frame(opcode, payload), at + size

File: ui/src/stream.ts
/**
 * The stream as the page opens it: where it is, and what arrives on it.
 *
 * The monitor is a page and a page cannot dial the mirror's port. What a
 * browser can open is a stream on the origin it is already on, which it does
 * by upgrading a request under the prefix the door strips before the mirror's
 * face sees it (ADR-0004 d.2, ADR-0004 d.3). On the other end of it is one
 * more client of the mirror's port, and every rule that port has reaches this
 * page because of it: the fan-out, the bound on how far behind a client may
 * fall, the outage that disconnects everyone (ADR-0007 d.2).
 *
 * **There is nothing on it but bytes.** No status, no progress, no link and no
 * protocol of the page's own (ADR-0008 d.2). What arrives is what the station
 * said, cut into lines at the newlines it wrote and stamped with the moment
 * they arrived — which is the only time there is, because the stream carries
 * none.
 *
 * **Nothing here reads a line.** What one means is the decoder's, which is a
 * pure function of a line and lands under its own ticket (ADR-0009), and
 * nothing is typed back up here at all: this is the monitor's downward half
 * (#4).
 */

/** The prefix the mirror's face answers under on this page's own origin, and
 *  the whole of what it claims there — the door strips it before the app sees
 *  a request (ADR-0004 d.2). */
export const FACE = "/dccex-usb";

/** Where the stream is. Opened by upgrading and fetched no other way: what it
 *  carries is what the station is saying now (`face.py`). */
export const STREAM_PATH = `${FACE}/stream`;

const HTTPS = "https:";
const WSS = "wss:";
const WS = "ws:";

/** How long the page waits before opening the stream again.
 *
 * Everything that ends a stream is something that ends: the mirror cuts off a
 * client that has fallen too far behind, an outage disconnects every client at
 * once, and the app restarting takes them all with it (ADR-0007 d.2). A
 * monitor that stayed dead until somebody reloaded the page would ask an
 * operator to reload it for a cable that was out for ten seconds.
 */
export const REOPEN_MS = 2000;

/**
 * Every byte a character, and no byte thrown away.
 *
 * The station's bytes ride as binary frames because a serial line promises no
 * UTF-8 (ADR-0007 d.5), and decoding them as UTF-8 on this end would put a
 * replacement character where the byte the station sent should be — the page
 * editing the conversation it is showing. Latin-1 is the decoding that is
 * total: one byte, one character, and a `<…>` message, which is ASCII,
 * unchanged. It is also what lets a read be decoded on its own, with no
 * character left half-arrived across two of them.
 */
const CHARACTERS = new TextDecoder("latin1");

const NEWLINE = "\n";
const RETURN = /\r+$/;

/** One line the station said, and the moment it arrived. */
export interface Said {
  /** When it arrived: the page's clock at the read that carried it. Two lines
   *  in one read carry the same stamp, which is what happened. */
  readonly at: Date;
  /** The line as the station said it, with the newline it ended off. */
  readonly line: string;
}

/** Where the stream is for the page `where` was read off.
 *
 * The page's own address with the scheme swapped and nothing else touched, so
 * a browser opens `wss://` from a page served over `https` and `ws://` from
 * one served over plain HTTP, on whatever host and port the page itself came
 * from. Nothing about the stream is a name or a port of its own (ADR-0004
 * d.3), and nothing in this page names a host: a page that did would work on
 * the machine it was written on and nowhere else.
 */
export function streamAt(where: Location): string {
  const at = new URL(STREAM_PATH, where.href);
  at.protocol = where.protocol === HTTPS ? WSS : WS;
  return at.href;
}

/** Fold `arrived` into `buffered` and take off every whole line, stamped `at`.
 *
 * Returns what is still a partial line, to be passed back as `buffered` next
 * time, and the lines that completed, in the order they did — the shape the
 * mirror's own readers have (`stream.py`, `framing.py`), and for the same
 * reason: bytes arrive in whatever chunks the network hands over, and a rule
 * that needed a line whole would be a rule about the network.
 *
 * A line that is empty once its newline is off is dropped. It is the `\r` of a
 * `\r\n` pair and a blank line the station wrote, and neither is something it
 * said.
 */
export function lines(
  buffered: string,
  arrived: string,
  at: Date,
): [string, Said[]] {
  const whole = buffered + arrived;
  const parts = whole.split(NEWLINE);
  const rest = parts.pop() ?? "";
  const said: Said[] = [];
  for (const part of parts) {
    const line = part.replace(RETURN, "");
    if (line) {
      said.push({ at, line });
    }
  }
  return [rest, said];
}

/**
 * The stream, open for as long as somebody is watching.
 *
 * It holds one socket, hands whole lines to whoever asked for them, and opens
 * another after `REOPEN_MS` when the one it has goes. It says nothing: the
 * page writes nothing up the stream under this ticket, and what the station
 * said while the stream was away is not in the one that follows — the mirror
 * keeps no history and there is nothing to ask for (ADR-0010).
 *
 * **A line the stream was cut off in the middle of is handed on as far as it
 * got.** Those bytes arrived, the newline that would have finished them never
 * will, and dropping them would be this page quietly losing what the station
 * said.
 */
export class Stream {
  readonly #said: (said: Said[]) => void;
  #socket: WebSocket | null = null;
  #reopening: ReturnType<typeof setTimeout> | null = null;
  #wanted = false;
  #partial = "";
  #partialAt = new Date();

  constructor(said: (said: Said[]) => void) {
    this.#said = said;
  }

  /** Open the stream, and keep one open until `close()`. */
  open(): void {
    this.#wanted = true;
    if (this.#socket === null && this.#reopening === null) {
      this.#dial();
    }
  }

  /** Let the stream go, and open no more of them: the monitor has left the
   *  page, so there is nobody to hand a line to. */
  close(): void {
    this.#wanted = false;
    if (this.#reopening !== null) {
      clearTimeout(this.#reopening);
      this.#reopening = null;
    }
    const socket = this.#socket;
    this.#socket = null;
    socket?.close();
  }

  #dial(): void {
    this.#reopening = null;
    this.#partial = "";
    const socket = new WebSocket(streamAt(window.location));
    socket.binaryType = "arraybuffer";
    socket.addEventListener("message", (said: MessageEvent<unknown>) => {
      this.#arrived(said.data);
    });
    // The socket going is the whole signal, as it is for every other client of
    // the mirror's port: a cut-off, an outage past its grace and the app going
    // down all reach a client as its connection ending (ADR-0007). An `error`
    // is followed by a `close`, so this is the one place either arrives at.
    socket.addEventListener("close", () => {
      this.#gone(socket);
    });
    this.#socket = socket;
  }

  #arrived(data: unknown): void {
    const text =
      typeof data === "string"
        ? data
        : data instanceof ArrayBuffer
          ? CHARACTERS.decode(data)
          : // A frame that is neither is not one the face sends: the station's
            // bytes ride as binary and `binaryType` is what this reads them as
            // (ADR-0007 d.5).
            "";
    const at = new Date();
    const [rest, said] = lines(this.#partial, text, at);
    this.#partial = rest;
    this.#partialAt = at;
    if (said.length > 0) {
      this.#said(said);
    }
  }

  #gone(socket: WebSocket): void {
    if (socket !== this.#socket) {
      return;
    }
    this.#socket = null;
    if (this.#partial) {
      this.#said([{ at: this.#partialAt, line: this.#partial }]);
      this.#partial = "";
    }
    if (this.#wanted) {
      this.#reopening = setTimeout(() => {
        this.#dial();
      }, REOPEN_MS);
    }
  }
}
